File: snippets/communication/consumer.py
from ndn.app import NDNApp
from ndn.encoding import Name

# Connect to the local forwarder over a Unix socket
app = NDNApp()

async def main():
    try:
        data_name, meta_info, content = await app.express_interest("/edu/ucla/cs/118/notes")

        # Received a Data packet
        print(f'Received Data Name: {Name.to_str(data_name)}')
        print(meta_info)
        print(bytes(content) if content else None)

    except InterestNack as e:
        # Received a Nack (negative acknowledgement)
        print(f'Nacked with reason={e.reason}')
    except InterestTimeout:
        # The Interest has timed out
        print(f'Timeout')
    except InterestCanceled:
        # Connection to the local NDN forwarder is broken
        print(f'Canceled')
    except ValidationFailure:
        # Security validation failed for the data
        print(f'Data failed to validate')

    finally:
        # Disconnect from the local forwarder
        app.shutdown()

if __name__ == '__main__':
    app.run_forever(after_start=main())

File: snippets/communication/consumer.cpp
#include <iostream>
#include <ndn-cxx/face.hpp>

int main(int argc, char** argv)
{
    // Create an Interest packet
    ndn::Interest interest("/edu/ucla/cs/118/notes");

    // Connect to the local forwarder over a Unix socket
    ndn::Face face;

    // Send the Interest packet and get back the Data packet
    face.expressInterest(interest,
                        [](const ndn::Interest&, const ndn::Data& data) {
                            // Received a Data packet reply
                            std::cout << data << std::endl;
                        },
                        [](const ndn::Interest&, const ndn::lp::Nack& nack) {
                            // Received a Nack (negative acknowledgement)
                            std::cout << "Nack received: " << nack.getReason() << std::endl;
                        },
                        [](const ndn::Interest&) {
                            // The Interest has timed out
                            std::cout << "Timeout" << std::endl;
                        });

    // Start face processing loop
    face.processEvents();
}